### PR TITLE
Add abstract serializer and deserializer

### PR DESCRIPTION
**Pull request by neithernut**
This PR introduces the abstract serializer and deserializer as well as an initial version of a message.

### CODE REVIEW

**Comment by neithernut** [comment]
Targets #161, obsoletes #155.

File: src/command/deserializer.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    size_t nbuf
+) {
+    // we don't have a message yet
+    self->is_ready = false;
+
+    // try to extract a message
+    ssize_t retval = self->deserialize(self, buf, nbuf);
+
+    // optimistic write
+    *msg = self->is_ready? self->buffer : NULL;
```
Comment (TheNeikos) [inline src/command/deserializer.c]: there should be a space before the `?`

File: src/command/deserializer.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * number of bytes read from the buffer.
+ * If the buffer didn't contain a complete message, the function should try to
+ * deserialize as much of the message as possible, updating the internal state
+ * and return the number of bytes read, not setting the `is_ready` flag.
+ * If the message was malformed in a non-recoverable way, the function should
+ * return `-EPROTO`.
+ * The function may also return a negative value if it encounters an internal
+ * error.
+ */
+typedef ssize_t (*ws_deserialiize_f)(struct ws_deserializer* self,
```
Comment (TheNeikos) [inline src/command/deserializer.h]: `deserialize` only takes one i

File: src/command/deserializer.c (hunk 1 above)
Comment (neithernut) [inline src/command/deserializer.c]: Mh... I'll add one, just to be sure.

File: src/command/deserializer.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with waysome. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "command/deserializer.h"
+
+/**
+ *
+ * Itnerface implementation
```
Comment (matthiasbeyer) [inline src/command/deserializer.c]: Typo.

**Comment by matthiasbeyer** [comment]
You add the serializer and the deserializer in the command directory, whereas we have `src/serialize` for this.

File: src/command/deserializer.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+#include <malloc.h>
+#include <stdbool.h>
+
+// Forward declarations
+struct ws_deserializer;
+struct ws_message;
+
+
+/**
+ * The deserialization  callback takes the deserializer itself, a pointer to
```
Comment (matthiasbeyer) [inline src/command/deserializer.h]: Doubled space.

**Comment by neithernut** [comment]
Argh, will fix this.

File: src/command/deserializer.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * the data which may be deserialized (`buf`) and the length of the buffer
+ * (`nbuf`).
+ *
+ * The function should deserialize a single message, update `nbuf` to the bytes
+ * processed as well as the messagei, set the `is_ready` flag and return the
```
Comment (matthiasbeyer) [inline src/command/deserializer.h]: Typo.

File: src/command/deserializer.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+/**
+ * Deserializer type
+ *
+ * A deserializer is initialized by a preparing function, which sets the
+ * callbacks and initializes the internal state to be used for parsing.
+ */
+struct ws_deserializer {
+    ws_deserialiize_f deserialize; //!< @protected deserialisation function
```
Comment (matthiasbeyer) [inline src/command/deserializer.h]: Typo from above here too.

File: src/command/message.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Message type
+ *
+ * Objects of this type represent messages both received and send to and fro
+ * modules.
+ *
+ * @extends ws_object
+ */
+struct ws_message {
+    struct ws_object obj; //!< @protected parent object
```
Comment (matthiasbeyer) [inline src/command/message.h]: Write "Base class" instead of "Parent object" please.

File: src/command/message.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+#ifndef __WS_COMMAND_MESSAGE_H__
+#define __WS_COMMAND_MESSAGE_H__
+
+#include "objects/object.h"
+
+/**
+ * Message type
+ *
+ * Objects of this type represent messages both received and send to and fro
```
Comment (matthiasbeyer) [inline src/command/message.h]: missing "m"

File: src/command/serializer.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+#include <malloc.h>
+
+// Forward declarations
+struct ws_serializer;
+struct ws_message;
+
+
+/**
+ * The serialization  callback takes the serializer itself, a pointer to
```
Comment (matthiasbeyer) [inline src/command/serializer.h]: Double space.

File: src/command/serializer.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+ * a target buffer to serialize to (`buf`), the length of the buffer in bytes
```
Comment (matthiasbeyer) [inline src/command/serializer.h]: The comma is misleading. There should be an "and" here.

File: src/command/serializer.h [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+
+/**
+ * Deserializer type
+ *
+ * A deserializer is initialized by a preparing function, which sets the
+ * callbacks and initializes the internal state to be used for parsing.
+ */
+struct ws_serializer {
+    ws_serialiize_f serialize; //!< serialization function
```
Comment (matthiasbeyer) [inline src/command/serializer.h]: Missing visibility tags for all members here.

File: src/command/serializer.h [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+ * return either the number of bytes written to that buffer or a negative error
+ * code.
+ * Note that in the case that the message could not be serialized because
+ * another message is still pending, `-EAGAIN` will be returned, indicating that
+ * the message may be serialized successfully later.
+ *
+ */
+ssize_t
+ws_serialize(
+    struct ws_serializer* self, //!< the serializer
```
Comment (matthiasbeyer) [inline src/command/serializer.h]: Missing visibility tags for all members here, too.

File: src/command/message.h (hunk 8 above)
Comment (neithernut) [inline src/command/message.h]: Nope, no "m" missing. It's supposed to be `fro`.

**Comment by neithernut** [comment]
Will move the message to objects.

File: src/command/message.h (hunk 8 above)
Comment (matthiasbeyer) [inline src/command/message.h]: Okay, I didn't know that word :sweat_smile:

File: src/serialize/deserializer.h [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+#include <malloc.h>
+#include <stdbool.h>
+
+// Forward declarations
+struct ws_deserializer;
+struct ws_message;
+
+
+/**
+ * The deserialization  callback takes the deserializer itself, a pointer to
```
Comment (matthiasbeyer) [inline src/serialize/deserializer.h]: Double space here.

File: src/serialize/serializer.h [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+#include <malloc.h>
+
+// Forward declarations
+struct ws_serializer;
+struct ws_message;
+
+
+/**
+ * The serialization  callback takes the serializer itself, a pointer to
```
Comment (matthiasbeyer) [inline src/serialize/serializer.h]: Double space here.

**Comment by neithernut** [comment]
Feel free to annotate more or ACK and merge.

**Comment by matthiasbeyer** [comment]
By now only one question regarding terminology: You add a `ws_message` type, but what about `ws_transaction` for example? Will this one be added in another PR or do we differ in our understanding of the terminology here?

Edit: If I'm right when saying that a `ws_message` is not related to a transaction that closely but more like "There is data from the client, lets call it 'message'"?

**Comment by neithernut** [comment]
If you look at the spec (section 4.3.2), you'll find that there are three types of messages either a client or waysome may send to each other: "transaction", "reply" and "event". A message of type "transaction" holds a `ws_transaction`.

So three types are derived from `ws_message`:
- `ws_message_transaction`
- `ws_message_reply`
- `ws_message_event`

**Comment by matthiasbeyer** [comment]
Okay, that was exactly what I was asking, I just did not remember that we called it "message" in the first place. :smile: 

ACK from me, @TheNeikos ?

**Comment by neithernut** [comment]
New stuff which I kind of forgot to include.

**Comment by neithernut** [comment]
Can this be merged? (I'll assume "yes" if I don't get a reply in the next 24h)

**Comment by matthiasbeyer** [comment]
Nope, don't!

File: src/objects/message.c [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+int
+ws_message_init(
+    struct ws_message* self,
+    size_t id
+) {
+    self->obj.id = &WS_OBJECT_TYPE_ID_MESSAGE;
+    self->id = id;
+
+    // what could possibly go wrong?
```
Comment (matthiasbeyer) [inline src/objects/message.c:67]: The call to `ws_object_init()` is missing completely here (above the `self->obj.id =` line, of course).

File: src/serialize/deserializer.c [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+    return retval;
+}
+
+void
+ws_deserializer_deinit(
+    struct ws_deserializer* self
+) {
+    ws_object_unref(&self->buffer->obj);
+    if (self->deinit) {
```
Comment (matthiasbeyer) [inline src/serialize/deserializer.c:67]: This looks as you would do a unref and use the object afterwards and got me the first time I looked at it. Are there any benefits for doing the unref before the other code? If not, I'd like to suggest you move the code below the if block.

If there are benefits, please add a blank line.

File: src/serialize/serializer.c [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+    return retval + offset;
+}
+
+void
+ws_serializer_deinit(
+    struct ws_serializer* self
+) {
+    ws_object_unref(&self->buffer->obj);
+    if (self->deinit) {
```
Comment (matthiasbeyer) [inline src/serialize/serializer.c:75]: Same as above here.

File: src/serialize/deserializer.c (hunk 16 above)
Comment (neithernut) [inline src/serialize/deserializer.c:67]: Check again, I unref **the buffer**.

File: src/serialize/serializer.c (hunk 17 above)
Comment (neithernut) [inline src/serialize/serializer.c:75]: Same as above.

File: src/serialize/deserializer.c (hunk 16 above)
Comment (matthiasbeyer) [inline src/serialize/deserializer.c:67]: Yes, of course. But the code is in one block, and that caused me to think that you unrefed the stuff you use in the next lines.

My point is not about how things _work_ but how things _look_. Please reread my comment from above carefully.

**Comment by matthiasbeyer** [comment]
ACK from me.

**Pull request merged**